src/Compilers/Core/Portable/PEWriter/MetadataWriter.cs

```diff
@@ -3065,6 +3066,9 @@ private void SubstituteFakeTokens(Blob blob, ImmutableArray<byte> methodBodyIL,
                                         case LiteralGreatestMethodDefinitionToken:
                                             token = GreatestMethodDefIndex;
                                             break;
+                                        case SourceDocumentIndex:
+                                            token = _dynamicAnalysisDataWriterOpt.GetOrAddDocument(((CommonPEModuleBuilder)module).GetSourceDocumentFromFakeToken((uint)(pseudoToken & 0x00ffffff)));
```

Comment:
This cast was the most worrisome part of this change set. I'll look into your idea for avoiding it.